src/pycell.rs

```diff
@@ -0,0 +1,678 @@
+}
+
+impl<T: PyClass + fmt::Debug> fmt::Debug for PyRef<'_, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&**self, f)
+    }
+}
+
+/// Wraps a mutable borrowed reference to a value in a `PyCell<T>`.
+///
+/// See the [`PyCell`](struct.PyCell.html) and [`PyRef`](struct.PyRef.html) documentations for more.
+pub struct PyRefMut<'p, T: PyClass> {
+    inner: &'p PyCellInner<T>,
+}
+
+impl<'p, T: PyClass> AsRef<T::BaseType> for PyRefMut<'p, T> {
+    fn as_ref(&self) -> &T::BaseType {
+        unsafe { &*self.inner.ob_base.get_ptr() }
+    }
+}
+
+impl<'p, T: PyClass> AsMut<T::BaseType> for PyRefMut<'p, T> {
+    fn as_mut(&mut self) -> &mut T::BaseType {
+        unsafe { &mut *self.inner.ob_base.get_ptr() }
+    }
+}
+
+impl<'p, T, U> PyRefMut<'p, T>
+where
+    T: PyClass + PyTypeInfo<BaseType = U, BaseLayout = PyCellInner<U>>,
+    U: PyClass,
+{
+    /// Get `PyRef<T::BaseType>`.
+    /// See  [`PyRef::into_super`](struct.PyRef.html#method.into_super) for more.
+    pub fn into_super(self) -> PyRefMut<'p, U> {
+        let PyRefMut { inner } = self;
+        std::mem::forget(self);
+        PyRefMut {
+            inner: &inner.ob_base,
+        }
+    }
+}
+
+impl<'p, T: PyClass> Deref for PyRefMut<'p, T> {
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &T {
+        unsafe { &*self.inner.get_ptr() }
+    }
+}
+
+impl<'p, T: PyClass> DerefMut for PyRefMut<'p, T> {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut T {
+        unsafe { &mut *self.inner.get_ptr() }
+    }
+}
+
+impl<'p, T: PyClass> Drop for PyRefMut<'p, T> {
+    fn drop(&mut self) {
+        self.inner.set_borrow_flag(BorrowFlag::UNUSED)
+    }
+}
+
+impl<'p, T: PyClass> FromPy<PyRefMut<'p, T>> for PyObject {
+    fn from_py(pyref: PyRefMut<'p, T>, py: Python<'_>) -> PyObject {
+        unsafe { PyObject::from_borrowed_ptr(py, pyref.inner.as_ptr()) }
+    }
+}
+
+impl<'a, T: PyClass> AsPyPointer for PyRefMut<'a, T> {
+    fn as_ptr(&self) -> *mut ffi::PyObject {
+        self.inner.as_ptr()
+    }
+}
+
+impl<'a, T: PyClass> std::convert::TryFrom<&'a PyCell<T>> for crate::PyRefMut<'a, T> {
+    type Error = PyBorrowMutError;
+    fn try_from(cell: &'a crate::PyCell<T>) -> Result<Self, Self::Error> {
+        cell.try_borrow_mut()
+    }
+}
+
+impl<T: PyClass + fmt::Debug> fmt::Debug for PyRefMut<'_, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&*(self.deref()), f)
+    }
+}
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+struct BorrowFlag(usize);
+
+impl BorrowFlag {
+    const UNUSED: BorrowFlag = BorrowFlag(0);
+    const HAS_MUTABLE_BORROW: BorrowFlag = BorrowFlag(usize::max_value());
+    const fn increment(self) -> Self {
+        Self(self.0 + 1)
+    }
+    const fn decrement(self) -> Self {
+        Self(self.0 - 1)
+    }
+}
+
+/// An error returned by [`PyCell::try_borrow`](struct.PyCell.html#method.try_borrow).
+pub struct PyBorrowError {
+    _private: (),
+}
```

Comment:
I wonder whether we should capture the Python pointer in here, so that we can report errors later:

```suggestion
pub struct PyBorrowError<'a> {
    ptr: &'a PyAny
}
```

Python users might like, for example, an error message more like:

```
Cannot borrow this object of type `Foo`, as it is already mutably borrowed.
```

Reply:
Hmm... :thinking: 
I think it's difficult since Error should be `'static`.

I'll rethink it when addressing #682